### PR TITLE
fix(api-client): socket reconnect after coming back from sleep

### PR DESCRIPTION
**Pull request by PatrykBuniX**
Once we've realised that the app was (**probably**) sleeping and during the possible "sleep time" the socket was disconnected, we will immediately reconnect. This should fix the bug where after coming back from sleep mode, messages are not synced immediately.

### CODE REVIEW

File: packages/api-client/src/tcp/ReconnectingWebsocket.ts [hunk 1]
```diff
@@ -173,16 +184,10 @@ export class ReconnectingWebsocket {
     return this.socket ? this.socket.readyState : WEBSOCKET_STATE.CLOSED;
   }
 
-  public disconnect(reason = 'Closed by client', keepClosed = true): void {
+  public disconnect(reason = 'Closed by client'): void {
     if (this.socket) {
       this.logger.info(`Disconnecting from WebSocket (reason: "${reason}")`);
-      // TODO: 'any' can be removed once this issue is resolved:
-      // https://github.com/pladaria/reconnecting-websocket/issues/44
-      (this.socket as any).close(CloseEventCode.NORMAL_CLOSURE, reason, {
```
Comment (PatrykBuniX) [inline packages/api-client/src/tcp/ReconnectingWebsocket.ts:181]: This method does not accept the 3rd (options) parameter any more.

File: packages/api-client/src/utils/BackFromSleepHandler/BackFromSleepHandler.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    const wasAsleep = currentTime > lastTime + TOLERANCE;
+
+    lastTime = currentTime;
+
+    if (isDisconnectedCallback && isDisconnectedCallback()) {
+      wasDisconnected = true;
+    }
+
+    if (wasAsleep) {
+      if (!isDisconnectedCallback || wasDisconnected) {
```
[the remark this answers is not in the document]
Comment (PatrykBuniX) [reply packages/api-client/src/utils/BackFromSleepHandler/BackFromSleepHandler.ts:64]: No, if the check callback is not provided at all we want to always run the callback

File: packages/api-client/src/tcp/ReconnectingWebsocket.ts [hunk 3]
```diff
@@ -82,6 +83,22 @@ export class ReconnectingWebsocket {
...
+     * To handle this, we need a custom approach to detect when the system goes to sleep and when it wakes up.
+     * **/
+    onBackFromSleep({
+      callback: () => {
+        if (this.socket) {
+          this.logger.debug('Back from sleep, reconnecting WebSocket');
+          this.socket?.reconnect();
+        }
+      },
+      isDisconnected: () => this.getState() === WEBSOCKET_STATE.CLOSED,
```
Comment (atomrc) [inline packages/api-client/src/tcp/ReconnectingWebsocket.ts:100]: Wasn't the ping/pong mechanism enough to detect that the connection was closed? 
Is there some kind of timer that runs to check if a message has been received in a certain amount of time?

Comment (PatrykBuniX) [reply packages/api-client/src/tcp/ReconnectingWebsocket.ts:100]: Yeah, we have the mechanism for that, but it's a ping every 20 seconds, which doesn't seem to be efficient enough to reconnect immediately after coming back from sleep. From what I was testing, it was the 20s that we had to wait for the weboscket to reconnect after coming back from sleep.

Comment (atomrc) [reply packages/api-client/src/tcp/ReconnectingWebsocket.ts:100]: Humm that makes sense, 20s is, indeed, pretty long time. 

Then your solution seems solid to me 😎

**Review by atomrc — APPROVED** [review]

**Pull request merged**